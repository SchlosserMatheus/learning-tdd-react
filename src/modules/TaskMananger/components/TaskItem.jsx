import React, {Component} from 'react';

import PropTypes from 'prop-types';

class TaskItem extends Component
{
    render() {

        const { checked, changeChecked, text} = this.props;
        const textStyle = (checked) ? {color: 'red'} : {color: 'black'};

        return (
            <div>
                <input type="checkbox" onChange={changeChecked} checked={checked}/>
                <span style={textStyle}>{text}</span>
            </div>
        );
    }
}

TaskItem.propTypes = {
    text: PropTypes.string.isRequired,
    checked: PropTypes.bool.isRequired,
    changeChecked: PropTypes.func.isRequired
};

export default TaskItem;