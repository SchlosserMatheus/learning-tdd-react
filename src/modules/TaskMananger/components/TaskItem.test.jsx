import React from 'react';
import {mount, shallow} from 'enzyme';
import TaskItem from "./TaskItem";
import sinon from "sinon";

describe("Testing TaskItem Component",() => {

    it("should render correctly",() => {
        const wrapper = shallow(<TaskItem text={'teste'} checked={false} changeChecked={ ()=> null} />);
        expect(wrapper.contains(
            <div>
                <input/>
                <span>teste</span>
            </div>
        ));
    });

    it('Should render input type checkbox', () => {
        const wrapper = shallow(<TaskItem text={'teste'} checked={false} changeChecked={ ()=> null} />);
        expect(wrapper.find('input[type="checkbox"][onChange][checked]').length).toEqual(1);
    });

    it('Should simulate event change of checkbox', () => {
        const changeChecked = sinon.spy();
        const wrapper = mount(<TaskItem changeChecked={changeChecked} text={'test'} checked={false}/>);
        wrapper.find('input').simulate('change');
        expect(changeChecked.callCount).toEqual(1);
    });

    it('Should render the text with red style when task is checked', () => {
        const wrapper = shallow(<TaskItem text={'teste'} checked={true} changeChecked={ ()=> null} />);
        const span = wrapper.find('span').first();
        if(span.length) expect(span.prop('style')).toHaveProperty('color', 'red');
        else{
            throw new Error(`Should have at least one span`);
        }
    });

    it('Should render the text with black style when task is not checked', () => {
        const wrapper = shallow(<TaskItem text={'teste'} checked={false} changeChecked={ ()=> null} />);
        const span = wrapper.find('span').first();
        if(span.length) expect(span.prop('style')).toHaveProperty('color', 'black');
        else{
            throw new Error(`Should have at least one span`);
        }
    });

    it('Check TaskItem snapshot version with prop checked true', () => {
        const wrapper = shallow(<TaskItem text={'teste'} checked={true} changeChecked={ ()=> null} />);
        expect(wrapper).toMatchSnapshot();
    });

    it('Check TaskItem snapshot version with prop checked false', () => {
        const wrapper = shallow(<TaskItem text={'teste'} checked={false} changeChecked={ ()=> null} />);
        expect(wrapper).toMatchSnapshot();
    });
});