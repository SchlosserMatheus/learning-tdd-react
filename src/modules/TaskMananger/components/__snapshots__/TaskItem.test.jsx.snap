// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing TaskItem Component Check TaskItem snapshot version with prop checked false 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <TaskItem
    changeChecked={[Function]}
    checked={false}
    text="teste"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <input
          checked={false}
          onChange={[Function]}
          type="checkbox"
        />,
        <span
          style={
            Object {
              "color": "black",
            }
          }
        >
          teste
        </span>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "checked": false,
          "onChange": [Function],
          "type": "checkbox",
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "teste",
          "style": Object {
            "color": "black",
          },
        },
        "ref": null,
        "rendered": "teste",
        "type": "span",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <input
            checked={false}
            onChange={[Function]}
            type="checkbox"
          />,
          <span
            style={
              Object {
                "color": "black",
              }
            }
          >
            teste
          </span>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "checked": false,
            "onChange": [Function],
            "type": "checkbox",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "teste",
            "style": Object {
              "color": "black",
            },
          },
          "ref": null,
          "rendered": "teste",
          "type": "span",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;

exports[`Testing TaskItem Component Check TaskItem snapshot version with prop checked true 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <TaskItem
    changeChecked={[Function]}
    checked={true}
    text="teste"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <input
          checked={true}
          onChange={[Function]}
          type="checkbox"
        />,
        <span
          style={
            Object {
              "color": "red",
            }
          }
        >
          teste
        </span>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "checked": true,
          "onChange": [Function],
          "type": "checkbox",
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "teste",
          "style": Object {
            "color": "red",
          },
        },
        "ref": null,
        "rendered": "teste",
        "type": "span",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <input
            checked={true}
            onChange={[Function]}
            type="checkbox"
          />,
          <span
            style={
              Object {
                "color": "red",
              }
            }
          >
            teste
          </span>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "checked": true,
            "onChange": [Function],
            "type": "checkbox",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "teste",
            "style": Object {
              "color": "red",
            },
          },
          "ref": null,
          "rendered": "teste",
          "type": "span",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
